File: src/discord/commands/public/moderate/utils/timeout.ts
import { res } from "#functions";
import { timeoutMenu, loadMenu, closeMenu } from "#menus";
import { createEmbed } from "@magicyan/discord";
import { User, Guild, InteractionReplyOptions, ChatInputCommandInteraction, ComponentType } from "discord.js";
import { filterUsers, showModal } from "./index.js";


function timeoutAction<R>(user: User, ids: string[], duration: string, reason: string, guild: Guild): R {
    let sucess: string[] = [];
    let failed: string[] = [];

    for (const id of ids) {
        const member = guild.members.cache.get(id);
        if (!member) {
            failed.push(id);
            continue;
        }
        member.timeout(parseInt(duration) * 1000, reason);
        sucess.push(id);
    }

    let description = `**Timeouted users:**\n${sucess.map(id => `<@${id}>`).join("\n")}`;
    if (failed.length > 0) {
        description = `${description}\n`
        description = `**Failed to timeout user(s):**\n${failed.map(id => `<@${id}>`).join("\n")}`
    }

    const embed = createEmbed({
        color: "Blue",
        author: {
            name: user.globalName || user.username,
            iconURL: user.avatarURL() || undefined
        },
        title: "**Officer Cui's panel**",
        thumbnail: "https://raw.githubusercontent.com/Cesio137/Bangboo/refs/heads/master/assets/avatar/Officer.png",
        description: description,
        footer: `Reason: ${reason}`
    });

    return ({
        flags: ["Ephemeral"],
        embeds: [embed],
        components: [],
    } satisfies InteractionReplyOptions) as R;
}

export async function timeoutCollector(interaction: ChatInputCommandInteraction<"cached">) {
    const { channel, guild, user } = interaction;
    if (!channel) {
        interaction.reply(
            res.danger("Invalid channel.")
        )
        return 
    }

    let ids: string[] = [];
    let duration: string = "";
    let timeout = true;
    let isOk = false;

    await interaction.deferReply({ flags: ["Ephemeral"] });
    let messageID = (await interaction.fetchReply()).id;
    interaction.editReply(timeoutMenu(user, ids, duration));

    const time = Date.now() + 300000;

    const userCollector = channel.createMessageComponentCollector(
        {
            componentType: ComponentType.UserSelect,
            filter: (componentInteraction) => componentInteraction.user.id === interaction.user.id && componentInteraction.message.id === messageID,
            time: time - Date.now(),
        }
    );

    const durationCollector = channel.createMessageComponentCollector(
        {
            componentType: ComponentType.StringSelect,
            filter: (componentInteraction) => componentInteraction.user.id === interaction.user.id && componentInteraction.message.id === messageID,
            time: time - Date.now(),
        }
    );

    const btnCollector = channel.createMessageComponentCollector(
        {
            componentType: ComponentType.Button,
            filter: (componentInteraction) => componentInteraction.user.id === interaction.user.id && componentInteraction.message.id === messageID,
            time: time - Date.now(),
        }
    );
    
    userCollector.on("collect", async function(i) {
        const { customId, guild, user } = i;
        if (!guild) { return }
        switch(customId) {
            case "mod/select-users":
                await i.update(loadMenu(user, "👥 **Filtering selected users...**"));
                ids = filterUsers(i.values, guild);
                await i.editReply(timeoutMenu(user, ids, duration));
                break;
        }
    });

    durationCollector.on("collect", async function(i) {
        const { customId, user } = i;
        switch(customId) {
            case "mod/select-duration":
                duration = i.values[0];
                await i.update(timeoutMenu(user, ids, duration))
                break;
        }
    });
    
    btnCollector.on("collect", async function(i) {
        const { customId, user } = i;
        switch(customId) {
            case "mod/btn-cancel":
                timeout = false;
                isOk = true;
                await i.update(closeMenu(i.user, timeout));
                userCollector.stop(); durationCollector.stop(); btnCollector.stop();
                break;

            case "mod/btn-confirm":
                const res = await showModal(i, time);
                if (res.isOk) {
                    i.editReply(timeoutAction(user, ids, duration, res.reason, guild));
                    timeout = false;
                    isOk = true;
                    userCollector.stop(); durationCollector.stop(); btnCollector.stop();
                }
                break;
        }
    });

    btnCollector.on("end", async function() {
        if (!isOk) interaction.editReply(closeMenu(interaction.user, timeout));
    });
}

File: src/menus/moderate/timeout.rs
use crate::helpers::*;
use serenity::all::{
    ButtonStyle, CreateActionRow, CreateButton, CreateComponent, CreateEmbed, CreateEmbedAuthor,
    CreateSelectMenu, CreateSelectMenuKind, CreateSelectMenuOption, User, UserId,
};
use std::borrow::Cow;

pub fn timeout_menu<'a>(
    user: &'a User,
    ids: &'a [UserId],
    duration: &'a str,
) -> (CreateEmbed<'a>, Vec<CreateComponent<'a>>) {
    let embed_author = CreateEmbedAuthor::new(user.global_name.as_ref().unwrap_or(&user.name))
        .icon_url(user.avatar_url().unwrap_or(user.default_avatar_url()));

    let embed = CreateEmbed::new()
        .color(str_hex_to_u32(&CONSTANTS.colors.royal))
        .author(embed_author)
        .title("**Officer Cui's panel**")
        .thumbnail("https://raw.githubusercontent.com/Cesio137/Bangboo/refs/heads/rust/assets/avatar/Officer.png")
        .description("🖱️ **Select user(s) and timeout duration!**");

    let user_row = CreateActionRow::SelectMenu(
        CreateSelectMenu::new(
            "mod/select-user",
            CreateSelectMenuKind::User {
                default_users: Some(Cow::Borrowed(ids)),
            },
        )
        .max_values(25)
        .min_values(0)
        .placeholder("Select user(s)"),
    );

    let duration_row = CreateActionRow::SelectMenu(
        CreateSelectMenu::new(
            "mod/select-duration",
            CreateSelectMenuKind::String {
                options: Cow::Owned(vec![
                    CreateSelectMenuOption::new("60 seconds", "60")
                        .default_selection(if duration == "60" { true } else { false }),
                    CreateSelectMenuOption::new("05 minutes", "300")
                        .default_selection(if duration == "300" { true } else { false }),
                    CreateSelectMenuOption::new("10 minutes", "600")
                        .default_selection(if duration == "600" { true } else { false }),
                    CreateSelectMenuOption::new("01 hour", "3600")
                        .default_selection(if duration == "3600" { true } else { false }),
                    CreateSelectMenuOption::new("01 weak", "604800")
                        .default_selection(if duration == "604800" { true } else { false }),
                ]),
            },
        )
        .min_values(1)
        .placeholder("Select duration"),
    );

    let confirm_row = CreateActionRow::Buttons(Cow::Owned(vec![
        CreateButton::new("mod/btn-cancel")
            .label("Cancel")
            .style(ButtonStyle::Danger),
        CreateButton::new("mod/btn-confirm")
            .label("Confirm")
            .style(ButtonStyle::Success)
            .disabled(ids.is_empty() || duration.is_empty()),
    ]));

    let components = vec![
        CreateComponent::ActionRow(user_row),
        CreateComponent::ActionRow(duration_row),
        CreateComponent::ActionRow(confirm_row),
    ];
    (embed, components)
}
